File: scala-advent-of-code/src/main/scala/advent/twenty_sixteen/Day06SignalsAndNoise.scala
package advent.twenty_sixteen

import advent.common.DailyProblem
import advent.utilities.FileUtilities

import scala.annotation.tailrec

class Day06SignalsAndNoise(filename: String) extends DailyProblem[String, String] {

  private val input = FileUtilities.readFile(filename)

  private def processInput(lines: List[String]): Map[Int, Map[Char, Int]] = {

    @tailrec
    def countCharacters(line: String, pos: Int, characters: Map[Int, Map[Char, Int]]): Map[Int, Map[Char, Int]] = {
      if (pos >= line.size) {
        characters
      } else {
        val c = line.charAt(pos)
        countCharacters(line, pos + 1, characters + (pos -> (characters(pos) + (c -> (characters(pos)(c) + 1)))))
      }
    }

    input.foldLeft(Map[Int, Map[Char, Int]]().withDefaultValue(Map[Char, Int]().withDefaultValue(0)))((acc, line) => countCharacters(line, 0, acc))
  }

  val result = processInput(input).toList.sortBy(entry => entry._1)

  override val part1Answer: String = result.map(entry => entry._2.toList.sortWith((l,r) => l._2  > r._2).head._1).mkString
  override val part2Answer: String = result.map(entry => entry._2.toList.sortWith((l,r) => l._2  < r._2).head._1).mkString
}

File: java-advent-of-code/src/main/java/advent/twenty_sixteen/Day06SignalsAndNoise.java
package advent.twenty_sixteen;

import advent.common.DailyProblem;
import advent.utilities.FileUtilities;
import advent.utilities.Parsers;

import java.util.*;

public class Day06SignalsAndNoise implements DailyProblem<String, String> {
    private static final Comparator<Map.Entry<Character, Integer>> SORTER = (o1, o2) -> Integer.compare(o2.getValue(), o1.getValue());

    private final String part1Answer;
    private final String part2Answer;

    public Day06SignalsAndNoise(String filename) {
        List<String> lines = FileUtilities.readLines(filename, Parsers::TO_STRING);

        Map<Integer, Map<Character, Integer>> counts = new HashMap<>();
        for (String line : lines) {
            for (int i = 0; i != line.length(); i++) {
                Map<Character, Integer> charMap = counts.getOrDefault(i, new HashMap<>());
                char c = line.charAt(i);
                charMap.put(c, charMap.getOrDefault(c, 0) + 1);
                counts.put(i, charMap);
            }
        }

        String signal1 = "";
        String signal2 = "";
        for (int i = 0; i != lines.get(0).length(); i++) {
            Map<Character, Integer> charCountMap = counts.get(i);
            LinkedList<Map.Entry<Character, Integer>> entries = new LinkedList<>(charCountMap.entrySet());

            Collections.sort(entries, SORTER);

            signal1 += entries.getFirst().getKey();
            signal2 += entries.getLast().getKey();
        }

        this.part1Answer = signal1;
        this.part2Answer = signal2;
    }

    @Override
    public String getPart1Answer() {
        return part1Answer;
    }

    @Override
    public String getPart2Answer() {
        return part2Answer;
    }
}
